contracts/core/Allo.sol

```diff
@@ -1,79 +1,368 @@
 // SPDX-License-Identifier: AGPL-3.0-only
 pragma solidity 0.8.19;
 
-import {Metadata} from "./libraries/Metadata.sol";
 import "@openzeppelin-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin-upgradeable/utils/MulticallUpgradeable.sol";
+import "@openzeppelin-upgradeable/token/ERC20/IERC20Upgradeable.sol";
 
-contract Allo is Initializable {
-    struct PoolData {
-        address identity;
-        address allocationStrategy;
-        address distributionStrategy;
+import "@solady/auth/Ownable.sol";
+
+import {Metadata} from "./libraries/Metadata.sol";
+import {Clone} from "./libraries/Clone.sol";
+import "../interfaces/IAllocationStrategy.sol";
+import "../interfaces/IDistributionStrategy.sol";
+import "./Registry.sol";
+
+contract Allo is Initializable, Ownable, MulticallUpgradeable {
+    /// @notice Custom errors
+    error UNAUTHORIZED();
+    error TRANSFER_FAILED();
+    error NOT_ENOUGH_FUNDS();
+    error STRATEGY_ALREADY_USED();
+    error NOT_APPROVED_STRATEGY();
+
+    /// @notice Struct to hold details of an Pool
+    struct Pool {
+        bytes32 identityId;
+        IAllocationStrategy allocationStrategy;
+        IDistributionStrategy distributionStrategy;
         Metadata metadata;
-        bool active;
     }
 
-    /**
-     * @notice Initializes the contract after an upgrade
-     * @dev In future deploys of the implementation, an higher version should be passed to reinitializer
-     */
-    function initialize() public reinitializer(1) {}
+    /// @notice Fee denominator
+    uint256 public constant FEE_DENOMINATOR = 1e18;
+
+    /// ==========================
+    /// === Storage Variables ====
+    /// ==========================
+
+    /// @notice Fee percentage
+    /// @dev 1e18 = 100%, 1e17 = 10%, 1e16 = 1%, 1e15 = 0.1%
+    uint256 public feePercentage;
+
+    /// @notice Incremental index
+    uint256 private _poolIndex;
+
+    /// @notice Nonce for cloning strategies
+    uint256 private _nonce;
+
+    /// @notice Allo treasury
+    address payable public treasury;
+
+    /// @notice Registry of pool creators
+    Registry public registry;
+
+    /// @notice Pool.id -> Pool
+    mapping(uint256 => Pool) public pools;
+
+    /// @notice Strategy -> bool
+    mapping(address => bool) public approvedStrategies;
```

Comment:
I'd probably call this `strategyLibrary` for clarity that other strategies are fine, these are just the ones we provide.

Reply:
i think we could rely on a indexing service ! making it iterable feels like an overkill